File: Go/Day03.go
package main

import AH "./adventhelper"

func stringIntersection(s1 string, s2 string) (intersection string) {
	set := make(map[rune]bool)

	for _, c := range s1 {
		set[c] = true
	}

	intersection = ""
	for _, c := range s2 {
		if _, ok := set[c]; ok {
			intersection = intersection + string(c)
		}
	}

	return
}

func score(bags []string) (total int) {
	total = 0

	for _, bag := range bags {
		half := len(bag) / 2
		lhs := bag[:half]
		rhs := bag[half:]

		intersection := stringIntersection(lhs, rhs)
		score := int(AH.FirstRune(intersection))
		if (score <= 90) {
			score -= 38
		} else {
			score -= 96
		}

		total += score
	}

	return
}

func assignBadges(bags []string) (total int) {
	total = 0
	for i := 0; i < len(bags); i = i + 3 {
		i1 := stringIntersection(bags[i], bags[i+1])
		i2 := stringIntersection(bags[i+2], i1)

		score := int(AH.FirstRune(i2))
		if (score <= 90) {
			score -= 38
		} else {
			score -= 96
		}
		total += score
	}

	return
}

func main() {
	bags, _ := AH.ReadStrFile("../input/input03.txt")

	AH.PrintSoln(3, score(bags), assignBadges(bags))

	return
}

File: C++/Day03.cpp
#include "AH.h"

namespace Day03
{
	
	std::string string_intersection(std::string s1, std::string s2) {
		std::sort(s1.begin(), s1.end());
		std::sort(s2.begin(), s2.end());

		std::string intersection;
		std::set_intersection(s1.begin(), s1.end(),
		                      s2.begin(), s2.end(),
		                      std::back_inserter(intersection));

		return intersection;
	}

	int score(const std::vector<std::string> bags)
	{
		int total = 0;

		for (auto & bag : bags)
		{
			const int half = bag.size() / 2;
			std::string lhs = bag.substr(0, half);
			std::string rhs = bag.substr(half, half);
			std::string i = string_intersection(lhs, rhs);

			int score = int(i.at(0));
			score -= (score <= 90) ? 38 : 96;

			total += score;
		}

		return total;
	}

	int assignBadges(const std::vector<std::string> ss)
	{
		int total = 0;
		for (size_t i = 0; i < ss.size(); i += 3)
		{
			std::string i1 = string_intersection(ss[i], ss[i+1]);
			std::string i2 = string_intersection(ss[i+2], i1);

			auto score = int(i2.at(0));
			score -= (score <= 90) ? 38 : 96;
			total += score;
		}

		return total;
	}

	int Run(const std::string& filename)
	{
		const auto rounds = AH::ReadTextFile(filename);
		
		AH::PrintSoln(3, score(rounds), assignBadges(rounds));

		return 0;
	}

}
